File: Week_07/G20200343030436/1122.数组的相对排序.java
/*
 * @lc app=leetcode.cn id=1122 lang=java
 *
 * [1122] 数组的相对排序
 */

// @lc code=start
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        int[] mapArray = new int[1001];
        int[] res = new int[arr1.length];
        int cnt = 0;
        for (int i = 0; i < arr1.length; i++) {
            mapArray[arr1[i]]++;
        }
        for (int i = 0; i < arr2.length; i++) {
            while (mapArray[arr2[i]] > 0) {
                res[cnt++] = arr2[i];
                mapArray[arr2[i]]--;
            }
        }
        for (int j = 0; j < 1001; j++) {
            while (mapArray[i] > 0) {
                res[cnt++] = i;
                mapArray[i]--;
            }
        }
        return res;
    }


       /**
 1  先计算arr1 中，每一个数字出现的次数的 dict
 2  遍历数组 arr2
 3 结合字典dict 获得每一值得次数，像结果res中append
 4 temp保存在arr1中，但不在arr2中的值，并使用sort排序
 5将result和 temp一起返回
 */
// func relativeSortArray(_ arr1: [Int], _ arr2: [Int]) -> [Int] {
//     var dict = [Int: Int]()
//     var result = [Int]()
//     for i in arr1 {
//         dict[i] = (dict[i] ?? 0) + 1//统计数字中 每个位置的值 出现的次数
//     }
//     for j in arr2 {
//         if let val = dict[j] {
//             for _ in 0 ..< val {
//                 result.append(j)//把子序列的值 枚举插入到 result数组里
//             }
//             dict.removeValue(forKey: j)//并从字典里删除这个数组2包含的 键值对
//         }
//     }
//     var simplaArray = [Int]()
//     if !dict.isEmpty {
//         for (key, value) in dict {
//             for _ in 0 ..< value {
//                 simplaArray.append(key)
//             }
//             dict.removeValue(forKey: key)
//         }
//     }
//     result.append(contentsOf: simplaArray.sorted())
//     return result
// }
}
// @lc code=end

File: Week_07/G20200447010018/1122.数组的相对排序.js
/*
 * @lc app=leetcode.cn id=1122 lang=javascript
 *
 * [1122] 数组的相对排序
 *
 * https://leetcode-cn.com/problems/relative-sort-array/description/
 *
 * algorithms
 * Easy (65.11%)
 * Likes:    36
 * Dislikes: 0
 * Total Accepted:    11.9K
 * Total Submissions: 18.1K
 * Testcase Example:  '[2,3,1,3,2,4,6,7,9,2,19]\n[2,1,4,3,9,6]'
 *
 * 给你两个数组，arr1 和 arr2，
 * 
 * 
 * arr2 中的元素各不相同
 * arr2 中的每个元素都出现在 arr1 中
 * 
 * 
 * 对 arr1 中的元素进行排序，使 arr1 中项的相对顺序和 arr2 中的相对顺序相同。未在 arr2 中出现过的元素需要按照升序放在 arr1
 * 的末尾。
 * 
 * 
 * 
 * 示例：
 * 
 * 输入：arr1 = [2,3,1,3,2,4,6,7,9,2,19], arr2 = [2,1,4,3,9,6]
 * 输出：[2,2,2,1,4,3,3,9,6,7,19]
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * arr1.length, arr2.length <= 1000
 * 0 <= arr1[i], arr2[i] <= 1000
 * arr2 中的元素 arr2[i] 各不相同
 * arr2 中的每个元素 arr2[i] 都出现在 arr1 中
 * 
 * 
 */

// @lc code=start
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function(arr1, arr2) {
    return arr1.sort((a, b) => {
        const idxA = arr2.indexOf(a);
        const idxB = arr2.indexOf(b);
        if (idxA === -1 && idxB === -1) {
            return a - b;
        } else if (idxA === -1) {
            return 1;
        } else if (idxB === -1) {
            return -1;
        } else {
            return idxA - idxB;
        }
    })
};
// @lc code=end

